integration_tests/test_pull_properties.py

```diff
@@ -86,3 +88,140 @@ def test_pull_global_build_packages_are_excluded(self):
 
         self.assertTrue(len(state.assets['build-packages']) == 0)
         self.assertNotIn('hello=2.10-1', state.assets['build-packages'])
+
+
+class GitAssetTrackingTestCase(testscenarios.WithScenarios,
+                               integration_tests.TestCase):
+
+    scenarios = [
+        ('plain', {
+            'part_name': 'git-part',
+            'expected': {}
+        }),
+        ('branch', {
+            'part_name': 'git-part-branch',
+            'expected': {
+                'branch': 'feature',
+            }
+        }),
+        ('tag', {
+            'part_name': 'git-part-tag',
+            'expected': {
+                'tag': 'feature-tag',
+            }
+        }),
+    ]
+
+    def test_pull_git(self):
+        project_dir = 'asset-tracking'
+
+        expected_commit = _source_helpers.create_git_repo('git-source')
+        self.run_snapcraft(['pull', self.part_name], project_dir)
+
+        state_file = os.path.join(
+            self.parts_dir, self.part_name, 'state', 'pull')
+        self.assertThat(state_file, FileExists())
+        with open(state_file) as f:
+            state = yaml.load(f)
+
+        self.assertIn('source-details', state.assets)
+
+        if 'branch' in self.expected:
+            self.assertEqual(self.expected['branch'],
+                             state.assets['source-details']['branch'])
+        elif 'tag' in self.expected:
+            self.assertEqual(self.expected['tag'],
+                             state.assets['source-details']['tag'])
+        else:
+            self.assertEqual(expected_commit,
```

Comment:
a comment in the test would be good. With the previous comment then we'd do something like

```python
if self.expected_details:
   self.assertThat(state.assets['source-details'][self.expected_details.key, Equals(self.expected_details.value))
else:
    # check the source-commit
    ...

```

`key` and `value` come from using named tuples.